File: code/1.py
import numpy as np
import matplotlib.pyplot as plt

a0 = 0.1 # MPa
f0 = 2.0 # MHz
w0 = 2 * np.pi * f0
T = 1 / f0 # microseconds

t = np.linspace(0, 2*T, 1000)
p_t = 2 * a0 * np.sin(3 * w0 * t) + a0 * np.cos(5 * w0 * t)

freqs_f0_multipliers = np.arange(-6, 7) 
freqs = freqs_f0_multipliers * f0

coeffs = np.zeros(len(freqs), dtype=complex)

coeffs[freqs_f0_multipliers == 3] = -0.1j
coeffs[freqs_f0_multipliers == -3] = 0.1j
coeffs[freqs_f0_multipliers == 5] = 0.05
coeffs[freqs_f0_multipliers == -5] = 0.05

amplitude = np.abs(coeffs)
phase = np.angle(coeffs)
real_part = np.real(coeffs)
imag_part = np.imag(coeffs)

fig = plt.figure(figsize=(12, 12))

plt.subplot(3, 1, 1)
plt.plot(t, p_t, 'b-', linewidth=2)
plt.title('Исходный акустический сигнал $p(t)$', fontsize=14)
plt.xlabel('Время $t$, мкс')
plt.ylabel('$p(t)$, МПа')
plt.grid(True)


plt.subplot(3, 2, 3)
plt.stem(freqs, amplitude)
plt.title('Амплитудный спектр $|p_T(f_n)|$')
plt.xlabel('Частота, МГц')
plt.grid(True)

plt.subplot(3, 2, 4)
plt.stem(freqs, phase)
plt.title('Фазовый спектр $arg(p_T(f_n))$')
plt.xlabel('Частота, МГц')
plt.ylabel('Радианы')
plt.grid(True)

# Действительная и мнимая части
plt.subplot(3, 2, 5)
plt.stem(freqs, real_part)
plt.title('Действительная часть $Re(p_T(f_n))$')
plt.xlabel('Частота, МГц')
plt.grid(True)

plt.subplot(3, 2, 6)
plt.stem(freqs, imag_part)
plt.title('Мнимая часть $Im(p_T(f_n))$')
plt.xlabel('Частота, МГц')
plt.grid(True)

plt.tight_layout()
for i, ax in enumerate(fig.axes):
    bbox = ax.get_position()
    print(f"График {i+1}: [{bbox.x0:.4f}, {bbox.y0:.4f}, {bbox.width:.4f}, {bbox.height:.4f}]")
plt.savefig('fig_1.png', dpi=300)

File: code/1.m
clear; clc; close all;

blue_color = [0.12, 0.47, 0.71];
red_color  = [0.84, 0.15, 0.16];
set(0, 'DefaultFigureColor', 'w');
set(0, 'DefaultAxesColor', 'w');
set(0, 'DefaultTextFontName', 'DejaVu Sans');
set(0, 'DefaultAxesFontName', 'DejaVu Sans');
set(0, 'DefaultAxesFontSize', 10);
set(0, 'DefaultTextFontSize', 12);
set(0, 'DefaultTextFontWeight', 'normal');
set(0, 'DefaultAxesTitleFontWeight', 'normal'); 
set(0, 'DefaultAxesTickDir', 'in');
set(0, 'DefaultAxesLineWidth', 1.5);
set(0, 'DefaultLineLineWidth', 1.7);
set(0, 'DefaultAxesTickLength', [0.015 0.025]);

set(0, 'DefaultAxesXGrid', 'on');
set(0, 'DefaultAxesYGrid', 'on');
set(0, 'DefaultAxesGridLineStyle', '-');
set(0, 'DefaultAxesGridColor', [0.6 0.6 0.6]);
set(0, 'DefaultAxesGridAlpha', 0.5);

a_0 = 0.1;
f_0 = 2.0;
w_0 = 2 * 3.14 * f_0;
T = 1 / f_0;
t = 0: 2*T: 1000;
p_t = 2 * a_0 * sin(3*w_0 * t) + a_0 * cos(5 * w_0 * t);
freqs_f0_multipliers = -6: 1: 6;
freqs = freqs_f0_multipliers * f_0;

coefficients = zeros(length(freqs), length(t));
coefficients(freqs_f0_multipliers == 3)  = -0.1i;
coefficients(freqs_f0_multipliers == -3) =  0.1i;
coefficients(freqs_f0_multipliers == 5)  =  0.05;
coefficients(freqs_f0_multipliers == -5) =  0.05;

c_amp = abs(coefficients);
c_phase = angle(coefficients);
c_real = real(coefficients);
c_im = imag(coefficients);

figure('Position', [110, 55, 1200, 1200]); 

axes('Position', [0.0667, 0.7066, 0.9208, 0.2606]); 
plot(t, p_t, 'b', 'LineWidth', 1.7);
title('Исходный акустический сигнал p(t)');
xlabel('Время t, мкс');
ylabel('p(t), МПа');
grid on;


axes('Position', [0.08, 0.38, 0.39, 0.22]); 
stem(freqs, c_amp, 'filled', 'Color', blue_color, 'MarkerFaceColor', blue_color);
hold on; plot(xlim, [0 0], 'Color', red_color, 'LineWidth', 1.7); hold off;
title('\bf a) \rm Амплитудный спектр |p_T(f_n)|');
xlabel('Частота f_n, МГц');
grid on;

axes('Position', [0.57, 0.38, 0.39, 0.22]); 
stem(freqs, c_phase, 'filled', 'Color', blue_color, 'MarkerFaceColor', blue_color);
hold on; plot(xlim, [0 0], 'Color', red_color, 'LineWidth', 1.7); hold off;
title('\bf b) \rm Фазовый спектр arg(p_T(f_n))')
xlabel('Частота f_n, МГц');
ylabel('Фаза p_T(f_n)');
grid on;

axes('Position', [0.08, 0.07, 0.39, 0.22]); 
stem(freqs, c_real, 'filled', 'Color', blue_color, 'MarkerFaceColor', blue_color);
hold on; plot(xlim, [0 0], 'Color', red_color, 'LineWidth', 1.7); hold off;
title('\bf c) \rm  Действительная часть спектра Re(p_T(f_n))')
xlabel('Частота f_n, МГц');
grid on;

axes('Position', [0.57, 0.07, 0.39, 0.22]); 
stem(freqs, c_im, 'filled', 'Color', blue_color, 'MarkerFaceColor', blue_color);
hold on; plot(xlim, [0 0], 'Color', red_color, 'LineWidth', 1.7); hold off;
title('\bf d) \rm  Мнимая часть спектра Im(p_T(f_n))')
xlabel('Частота f_n, МГц');
grid on;


print(gcf, 'fig_1_m.png', '-dpng', '-r300');
